### ISSUE TITLE
Why did RabbitMQ eat all my memory?

### ISSUE THREAD

**Issue by gerhard** [rabbitmq/tgir#5]
Regardless how much [guidance](https://www.rabbitmq.com/memory-use.html) & [effort](https://grafana.gcp.rabbitmq.com/dashboard/snapshot/wM5JcgR9oQToU4CR54IbOq1NtAsa6jvu?orgId=1) we have put into explaining and reasoning about RabbitMQ's use of system memory, we keep finding ourselves in these types of situations:

## @Farkie via [https://rabbitmq.slack.com](https://rabbitmq.slack.com/archives/C1EDN83PA/p1580423414188300)

Hey everyone - I have a decent size rabbitmq server with 61G of RAM, sitting on a 0.4 watermark default, which gives a 24GB High watermark. It's been up for about 7 days now, and memory is hovering around 23G/24G. Our monitoring is picking this up as super high usage (which I guess is correct), but if I use the cli to get stats out, I see  `allocated_unused: 19.2819 gb (75.78%)` which makes me think this isn't a problem at all? Anything I can do to either drop the RAM usage or get it show the real usage not just allocated? 

**I don't think there's harm in leaving this alone...**

> hehe you don't have pagerduty screaming at you 😉 

**Restart it?**

> Last time I restarted the node, it crashed on shutdown and I lost all persistent messages 😱 
> it got stuck on rebuilding indexes
> I think it kept getting hit by OOM
> so it was in an endless loop
> I even doubled the RAM to 122G and tried to start up

**What are you running?**

> Rabbit v3.7.2 with Erlang v20.1.7

**Can we see your RabbitMQ Overview, rabbitmq_top & memory breakdown?**

![slack-imgs](https://user-images.githubusercontent.com/3342/73645524-5fc0b380-466f-11ea-9c3f-ac943e7f9608.png)

![slack-imgs](https://user-images.githubusercontent.com/3342/73642203-2d13bc80-4669-11ea-8014-8e8f58b87b39.png)

![slack-imgs-1](https://user-images.githubusercontent.com/3342/73642265-4e74a880-4669-11ea-8f28-60e2808570f3.png)

![queue-properties](https://user-images.githubusercontent.com/3342/73664618-9e6a6400-4697-11ea-8c31-fd6381de93bc.png)

![message-properties](https://user-images.githubusercontent.com/3342/73664630-a0342780-4697-11ea-974b-2fe8cc649f2d.png)

```
[root@liverabbitone centos]# rabbitmqctl eval 'recon:bin_leak(10).'
[{<9202.2247.0>,-22841,
  [{current_function,{gen_server2,process_next_msg,1}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.511.0>,-15892,
  [channel_queue_exchange_metrics_metrics_collector,
   {current_function,{gen_server,loop,7}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.2667.0>,-15865,
  [{current_function,{gen_server2,process_next_msg,1}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.2635.0>,-15865,
  [{current_function,{gen_server2,process_next_msg,1}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.2127.0>,-2160,
  [{current_function,{gen_server2,process_next_msg,1}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.513.0>,-1443,
  [channel_exchange_metrics_metrics_collector,
   {current_function,{gen_server,loop,7}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.12165.991>,-955,
  [rabbit_mgmt_db_cache_connections,
   {current_function,{gen_server,loop,7}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.2043.0>,-558,
  [{current_function,{gen_server2,process_next_msg,1}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.3353.0>,-548,
  [{current_function,{gen_server2,process_next_msg,1}},
   {initial_call,{proc_lib,init_p,5}}]},
 {<9202.2059.0>,-522,
  [{current_function,{gen_server2,process_next_msg,1}},
   {initial_call,{proc_lib,init_p,5}}]}]
```

**Well, maybe your best option would be to stand up the latest version and re-direct your traffic there. Then use the shovel to move existing messages. https://www.rabbitmq.com/blue-green-upgrade.html**

> oh that's cool, thanks for that 🙂 

@Farkie can you please add a RabbitMQ Overview screenshot from your environment to this issue? I would like to see the global counts. A sanitized `rabbitmqctl report` attached as a file would be champion.

Required for this issue to be taken seriously:
![Strip-Chrome-bouffe-de-la-RAM-english650-final1](https://user-images.githubusercontent.com/3342/73642829-508b3700-466a-11ea-9132-a922ad6af28f.jpg)

**Comment by gerhard** [comment]
FWIW, this erlang-questions thread is most relevant: http://erlang.org/pipermail/erlang-questions/2020-January/099077.html

**Comment by gerhard** [comment]
## Spec

* 1 node RabbitMQ v3.7.2 on Erlang/OTP v20.1.7
* 8 vCPU
* 60GB RAM

## Topology

* topic exchanges
* durable queues
* 500 queues with `{"x-expires",2592000000},{"x-message-ttl",2592000000}` (30 days)
  * not all queues have clients
  * not all queues have TTL
  * 2 queues have 12mil messages
  * 5 queues have 3mil messages
  * 3 queues have 1mil messages
  * 14 queues have 20k - 600k messages
* msg size 26KB
* 150 connections
* clients use autoack, 50 prefetch

## RabbitMQ Config Highlights

* `{disk_free_limit,4000000000}`
* `{vm_memory_high_watermark,0.4}`
* `{vm_memory_high_watermark_paging_ratio,0.6}`

## Links

* https://github.com/rabbitmq/rabbitmq-server/issues/1284#issuecomment-355959203
* https://www.rabbitmq.com/lazy-queues.html#caveats-limitations

**Comment by michaelklishin** [comment]
Let's not turn this into a support forum. We have had a number of highly relevant metrics in place for a while. The fact that many are unaware of them or go into production without monitoring is what a new episode should address. If there are specific metrics that could help there are missing, they deserve to have their own specific issues.

**Comment by gerhard** [comment]
If it wasn't for:

* node using almost as much memory as the `vm_memory_high_watermark` during TGIR S01E01, then using 1/3 after reboot (exactly the same workload)
* [this recent erlang-questions thread](http://erlang.org/pipermail/erlang-questions/2020-January/099077.html)
* outstanding ETS memory allocator optimisations for Quorum Queues in v3.8.x
* discussions with @carlhoerberg during RabbitMQ Summit 2018 about this very topic
* the next iteration of [How can we increase multiblock carrier utilization for binary_alloc?](http://erlang.org/pipermail/erlang-questions/2018-May/095327.html)

then I wouldn't feel as strongly about this.

[I have been on this journey long enough](https://github.com/rabbitmq/rabbitmq-server/issues/1223) to know that it's time for the next small step that will make RabbitMQ's relationship with memory that little bit clearer and, hopefully, better.

**Comment by gerhard** [comment]
Closed via #7
